### ISSUE TITLE
Suggestion: disallow use before definition

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#21]
The compiler should issue an error when code uses values before they could possibly be initialized.

```TypeScript
// Error, 'Derived' declaration must be after 'Base'
class Derived extends Base { }
class Base { }
```

**Comment by vladima** [comment]
Ok, what rules we would like to implement as a part of this issue besides 'base class should be lexically defined before the derived class'?

**Comment by RyanCavanaugh** [comment]
Disallow forward references to internal module members, e.g.
```ts
var x = M.fn(); // Should error
module M {
	export function fn() {}
}
```

Disallow forward references to enum members
```ts
var x = E.A; // Should error
enum E { A }
```

**Comment by NoelAbrahams** [comment]
IMO the scope of this issue is rather limited as it stands because people do not normally define all their code in a single file: the base class is more likely to exist in a separate file to the derived class.

I suggest the following from @sparecycles should also be part of the resolution to this issue:

> Then we shouldn't change the order but should we at least (have the option to) warn the user that the order the compiler uses is probably wrong?

The "order the compiler uses" should include the order specified in `tsconfig`.

**Comment by sparecycles** [comment]
When the base class and derived classes are in the same file, the problem is not as bad: the program will crash on startup and the programmer will change the order in that one file and the problem will be *fixed*.

The multiple file case is where the compiler should warn when concatenating multiple files in a dubious order, because that order can change for subtle reasons.

Consider the case where there is a base class and multiple derived classes, all in separate files.  The base class uses some of the derived classes in its implementation, and so it references them, but it still needs to be placed first in the output.  Similarly, all the derived classes need to reference the base class.

Well, there is no problem with mutually referenced files, if A.ts mutually references B.ts and X.ts includes A.ts, then the output order will be [B,A,X], and if it references B.ts the order will be [A,B,X].  (But only one of these orders might work at runtime.) This makes things fragile, as compilation will succeed equally well if either B or A is referenced.

My solution for my base-derived class system problem: add an index.ts for my class hierarchy and include, in this file, all the derived classes, followed by the base class.  This guaranteed that the output would put the base class first. (completely counter-intuitive!).  I found that if I directly referenced the files I wanted it would end up generating the base class after the derived one.

The compiler warning will be very nice, but it would also be great to be able to flag one of the references in a mutual-reference scenario as emit-ordering, and the other is just for pulling in declarations.  Mutual emit-ordering references would be an error.

( I currently have this implemented for automatically generating the list of .js includes in our Visual Studio/Typescript project since we're using multi-file output (easier to debug).  But the code is in C# as an inline task. If there is interest I will ask if I can share it.  It's basically two runs of Tarjan's CC algorithm. )

Both warning when emit-order is wrong and stabilizing emit-order with an explicit directive would go a long way toward making typescript a viable language for large projects... yes?

**Comment by xwipeoutx** [comment]
I run into this issue fairly frequently with my rather small code base (around 80 .ts files).  Ideally, I'd like to not have any `<reference>` tags at the top of any of my files, and the compiler can work it all out for me.  

My app has only 1 file that instantiates classes and executes the application (my composition root), a few files that add extensions (eg. adding `Array.prototype.distinct`) and the rest are just class/interface definitions.

In this case, most of the code is fair game for reordering, and shouldn't require manually `<reference>` definitions to get right. I see class definitions as being fair game for any compiler reordering, and should be shunted up to the top of the combined output, while the rest of the statements can preserve ordering as it was when input.

Would a compiler flag `--looseSorting` be possible? It seems like a fairly sought after feature.

**Comment by yuit** [comment]
In emit class-declaration in ES6, we do static property assignment after class-declaration. This makes referring to class static property in computed property name becomes use-before-definition. 

Emitted JS:
```
class C {
    [C.p] () {}  // Use before definition
    [C.p+ C.e]() {}  // Use before definition
    [D.f] () {}  // Use before definition
}
C.p = 10;
C.e = 20;

class D {
}
D.f = "hi";
```
We only want to warn about this error in two cases: computed-property names refers to its class static property, or refer to other class, that defined below it, property.

**Comment by DanielRosenwasser** [comment]
The trivial example we were playing with today, just to include in any tests:

```TypeScript
function f() {
    function g() {
        i = 10;
    }

    let i = 20;
    g();
}
```

It'd be good to get the permutations of uses/definitions of `g` around `i`.

**Comment by jvilk** [comment]
Don't forget to think about functions defined within block scope. That's undefined behavior as per the JavaScript standard, and I know at least Firefox and Chrome disagree in their implementation.

e.g.:

```{typescript}
function f() {
    if (true) {
        g(); // iirc, g executes in Chrome, and is undefined in Firefox
        function g() {
        }
        g(); // works in both browsers
    }
}
```

**Comment by danquirk** [comment]
Tracked by https://github.com/Microsoft/TypeScript/issues/2854 now.

**Comment by BrainCrumbz** [comment]
Just to mention that we've just being bitten by this today, and it took us some time to figure out what was going on. 

TypeScript v1.8.10, Webpack-based build, both base and derived class defined in same file, but (apparently) in the wrong order, no compile errors nor warnings, and even if source maps are working, the error call stack was pointing to a highly unuseful location (the end of another class importing the derived one).

Not going through the whole discussion, but as a first aid it seems like a compiler warning would help. Just our 2¢

**Comment by MrGuardian** [comment]
I find it ridiculous that TS does not support this feature out of the box. The confusion it causes is similar as using standard JS. Also, virtual methods, anyone?

**Comment by RyanCavanaugh** [comment]
@MrGuardian the repro described in the OP has been fixed. Perhaps you can clarify in a new issue or existing issue that better describes the problem you're having?

**Comment by Spongman** [comment]
here's another two cases that IMO should be errors:

````
class Test
{
    _b = this._a; // undefined, no error/warning
    _a = 3;

    static _B = Test._A; // undefined, no error/warning
    static _A = 3;
    
    method()
    {
        let a = b; // Block-scoped variable 'b' used before its declaration
        let b = 3;
    }
}
````

**Comment by RyanCavanaugh** [comment]
@Spongman can you log that in a separate issue please? Thanks!

**Comment by Spongman** [comment]
#12673
